File: library/src/main/java/jp/daneko/android/inappv3/model/ProductBaseInfo.java
package jp.daneko.android.inappv3.model;

import fj.Ord;
import fj.Ordering;

import lombok.NonNull;
import lombok.Value;

/**
 */
@Value
public class ProductBaseInfo implements Comparable<ProductBaseInfo> {
    @NonNull
    String id;
    @NonNull
    BillingType type;

    @Override
    public int compareTo(ProductBaseInfo another) {
        final int x = type.compareTo(another.getType());
        return x == 0 ? id.compareTo(another.getId()) : x;
    }

    public static Ord<ProductBaseInfo> ord = Ord.<ProductBaseInfo>ord(
            a1 -> a2 -> {
                final int x = a1.compareTo(a2);
                return x < 0 ? Ordering.LT : x == 0 ? Ordering.EQ : Ordering.GT;
            });
}

File: library/src/main/kotlin/com/github/daneko/android/iab/model/ProductBaseInfo.kt
package com.github.daneko.android.iab.model

import fj.F2Functions
import fj.Ord
import fj.Ordering


/**
 */
class ProductBaseInfo(internal val id: String, internal val type: BillingType) : Comparable<ProductBaseInfo> {

    override fun compareTo(other: ProductBaseInfo): Int {
        val x = type.compareTo(other.type)
        return if (x == 0) id.compareTo(other.id) else x
    }

    companion object {
        @JvmStatic
        val ord: Ord<ProductBaseInfo> = Ord.ord(F2Functions.curry { a: ProductBaseInfo, b: ProductBaseInfo ->
            val x = a.compareTo(b)
            if (x < 0) Ordering.LT else if (x == 0) Ordering.EQ else Ordering.GT
        })
    }
}
